pkg/timelock/timelock.go

```diff
@@ -240,54 +258,73 @@ func (tw *Worker) Listen(ctx context.Context) error {
 				// Check if the error is not nil, because sub.Unsubscribe will
 				// signal the channel sub.Err() to close it, leading to false nil errors.
 				if err != nil {
-					tw.logger.Info().Msgf("subscription: %s", err.Error())
+					tw.logger.Warn().Msgf("subscription error: %s", err.Error())
 					SetReadyStatus(HealthStatusError)
-					loop = false
+					sub.Unsubscribe()
+
+					success := false
+					for try := range maxSubRetries {
+						tw.logger.Warn().Msgf("trying to re-create subscription: %v/%v retry.", try+1, maxSubRetries)
+						sub, err = tw.ethClient.SubscribeFilterLogs(ctx, query, logCh)
+						if err == nil {
+							tw.logger.Info().Msg("subscription successfully recreated.")
+							SetReadyStatus(HealthStatusOK)
+							success = true
+							break
+						}
+
+						time.Sleep(time.Second * time.Duration(try))
+					}
+
+					if !success {
+						tw.logger.Error().Msg("failed to recreate subscription after retries: shutting down timelock-worker.")
+						break MainLoop
+					}
 				}
 
 			case signal := <-stopCh:
 				tw.logger.Info().Msgf("received OS signal %s", signal)
 				SetReadyStatus(HealthStatusError)
-				loop = false
+				break MainLoop
 			}
 		}
 		wg.Done()
 	}()
-	wg.Wait()
-
-	// Close in this specific order to avoid runtime panics,
-	// or memory leaks.
-	defer close(sigCh)
-	defer close(stopCh)
-	defer close(logCh)
-	defer sub.Unsubscribe()
-	defer ctx.Done()
 
-	tw.dumpOperationStore(time.Now)
+	wg.Wait()
```

Comment:
goes to my point about combining the go routine that reads the finite result from filter query and the subscription into one adds confusion into this method.  really this wait could be in Listen.